File: array/1768. Merge Strings Alternately.py
### Algorithm ###
"""
while any not empty:
    if word1 exist: 
        result += word1[0]
        word1 = word1[1:]
    if word2 exist:
        result += word2[0]
        word2 = word2[1:]
"""

### Complexity ###
"""
Time complexity
    A for loop -> O(n), n is max(len(word1), len(word2))

Space complexity
    result store length n+m string
"""

### Implementation ###

def mergeAlternately(word1, word2):
    """
    :type word1: str
    :type word2: str
    :rtype: str
    """
    
    result = ""
    # preallocate
    result = chr(0) * len(word1 + word2)

    while word1 or word2:
        if word1:
            result += word1[0]
            word1 = word1[1:]
        if word2:
            result += word2[0]
            word2 = word2[1:]
    
    return result

### Test Function ###

def testMergeAlternately(word1, word2, want):

    result = mergeAlternately(word1=word1, word2=word2)
    
    return "pass" if result == want else f"want {want}, got {result}"


# testcases
print("case 1: ", testMergeAlternately("abc", "pqr", "apbqcr"))
print("case 2: ", testMergeAlternately("ab", "pqrs", "apbqrs"))
print("case 3: ", testMergeAlternately("abcd", "pq", "apbqcd"))

File: array/1768. Merge Strings Alternately.go
package array

import "strings"

/*
	// Use strings.Split() and concatenation
	func MergeAlternately(word1 string, word2 string) string {
		result := ""

		splited1 := strings.Split(word1, "")
		splited2 := strings.Split(word2, "")

		for {
			if len(splited1) == 0 && len(splited2) == 0 {
				break
			}

			if len(splited1) > 0 {
				result += splited1[0]
				splited1 = splited1[1:]
			}

			if len(splited2) > 0 {
				result += splited2[0]
				splited2 = splited2[1:]
			}
		}

		return result
	}

	// Time complexity: O(n)
	// Space complexity: O(n) for result, O(n) for splited1, and O(n) for splited2
*/

// Use strings.Builder
func MergeAlternately(word1 string, word2 string) string {
	var result strings.Builder
	result.Grow(len(word1 + word2)) // ensure the result's capacity

	for i := 0; i < len(word1+word2); i++ {

		if i > len(word1) && i > len(word2) {
			break
		}

		if i < len(word1) {
			result.WriteByte(word1[i])
		}

		if i < len(word2) {
			result.WriteByte(word2[i])
		}
	}

	return result.String()
}

// Time complexity: O(n)
// Space complexity: O(n) for result
